tensorflow/python/training/adam.py

```diff
@@ -174,7 +175,51 @@ def _resource_apply_dense(self, grad, var):
         math_ops.cast(self._epsilon_t, grad.dtype.base_dtype),
         grad, use_locking=self._use_locking)
 
-  def _apply_sparse_shared(self, grad, var, indices, scatter_add):
+  class _ScatterOpWrapper(object):
+    """Wraps necessary scatter ops for sparse tensors."""
+
+    def __init__(self, use_locking=False):
+      self._use_locking = use_locking
+
+    def add(self, sparse_tensor, index, delta):
+      return state_ops.scatter_add(sparse_tensor, index, delta,
```

Comment:
Rename sparse_tensor to variable?

Also do variable.scatter_add(...) instead of state_ops.scatter_add(...) as that is the syntax for tf 2.0.

Doing variable.scatter_add will also obviate the need for a separate resource variable implementation.